keps/sig-node/4112-passdown-resources-to-cri/README.md

```diff
@@ -0,0 +1,1300 @@
++
++    // Type of the container
++    ContainerType type= 2;
++
++    // Kubernetes resource spec of the container
++    KubernetesResources kubernetes_resources = 3;
++
++    // Mounts for the container.
++    repeated Mount mounts = 4;
++
++    // Devices for the container.
++    repeated Device devices = 5;
++
++    // CDI devices for the container.
++    repeated CDIDevice CDI_devices = 6;
++}
+
++enum ContainerType {
++    INIT_CONTAINER    = 0;
++    SIDECAR_CONTAINER = 1;
++    REGULAR_CONTAINER = 2;
++}
+```
+
+The Pod-level resources enhancement [KEP-2837][kep-2837]
+([alpha PR][kep-2837-alpha-pr]) proposes to add new Pod-level resource
+requirements fields to the PodSpec. This information will be be added to the
+PodResourceConfig message, similar to the container resource information.
+
+```diff
+ message PodResourceConfig {
+     repeated ContainerResourceConfig containers = 1;
++
++    // Kubernetes resource spec of the pod-level resource requirements.
++    KubernetesResources kubernetes_resources = 2;
+ }
+```
+
+The implementation if adding the KubernetesResources field to the
+PodResourceConfig is synced with [KEP-2837][kep-2837].
+
+#### CreateContainer
+
+The ContainerConfig message (used in CreateContainer request) is extended to
+contain unmodified resource requests from the PodSpec.
+
+```diff
++import "k8s.io/apimachinery/pkg/api/resource/generated.proto";
+
+ message ContainerConfig {
+ 
+ ...
+ 
+     // Configuration specific to Windows containers.
+     WindowsContainerConfig windows = 16;
+ 
+     // CDI devices for the container.
+     repeated CDIDevice CDI_devices = 17;
++
++    // Kubernetes resource spec of the container
++    KubernetesResources kubernetes_resources = 18;
+ }
+ 
++// KubernetesResources contains the resource requests and limits as specified
++// in the Kubernetes core API ResourceRequirements.
++message KubernetesResources {
++    // Requests and limits from the Kubernetes container config.
++    map<string, k8s.io.apimachinery.pkg.api.resource.Quantity> requests = 1;
++    map<string, k8s.io.apimachinery.pkg.api.resource.Quantity> limits = 2;
++}
+```
+
+Note that mounts, devices, CDI devices are part of the ContainerConfig message
+but are left out of the diff snippet above.
+
+Including the KubernetesResources in the ContainerConfig message serves
+multiple purposes:
+
+1. Catch changes that happen between pod sandbox creation and container
+   creation. For example, in-place pod updates might change the container
+   before it was created.
+2. Catch changes that happen over container restarts in in-place pod update
+   scenarios
+3. Consistency/completeness. Have enough information to make consistent action
+   based only on information present in this rpc caal.
+
+The resources (mounts, devices, CDI devices, Kubernetes resources) in the
+CreateContainer request should be identical to what was (pre-)informed in the
+RunPodSandbox request. If they are different, the CRI runtime may fail the
```

Comment:
how would we separate cases when VM was restarted and we don't really know what was originally passed to the sandbox creation any longer? I do not understand this consistency check argument

Reply:
Yes, this paragraph seems outdated, especially in the light of the in-place pod updates. The paragraph was added as a response to some review feedback (I just couldn't find that comment now)

How about re-phrasing it for example:

```
In some usage scenarios the container creation may fail if
the resources (mounts, devices, CDI devices, Kubernetes resources) in the       
CreateContainer request do not match what was (pre-)informed in the   
RunPodSandbox request. This may be the case e.g. in when 
changes are not allowed after a VM-based Pod has been created.
```

WDYT @SergeyKanzhelev ?

Reply:
I believe he means the original pod metadata plus any pod updates must be applied to the pod before the container is created.